### PR TITLE
[SPARK-15646] [SQL] When spark.sql.hive.convertCTAS is true, the conversion rule needs to respect TEXTFILE/SEQUENCEFILE format and the user-defined location

### PR DESCRIPTION
**Pull request by yhuai**
## What changes were proposed in this pull request?

When `spark.sql.hive.convertCTAS` is true, for a CTAS statement, we will create a data source table using the default source (i.e. parquet) if the CTAS does not specify any Hive storage format. However, there are two issues with this conversion logic.
1. First, we determine if a CTAS statement defines storage format by checking the serde. However, TEXTFILE/SEQUENCEFILE does not have a default serde. When we do the check, we have not set the default serde. So, a query like `CREATE TABLE abc STORED AS TEXTFILE AS SELECT ...` actually creates a data source parquet table.
2. In the conversion logic, we are ignoring the user-specified location.

This PR fixes the above two issues. 

Also, this PR makes the parser throws an exception when a CTAS statement has a PARTITIONED BY clause. This change is made because Hive's syntax does not allow it and our current implementation actually does not work for this case (the insert operation always throws an exception because the insertion does not pick up the partitioning info).
## How was this patch tested?

I am adding new tests in SQLQuerySuite and HiveDDLCommandSuite.

### CODE REVIEW

**Comment by yhuai** [comment]
@ericl @andrewor14 @liancheng Can you review this PR?

**Comment by yhuai** [comment]
test this please

**Comment by yhuai** [comment]
OK. External related changes will be handled by another PR.

**Comment by yhuai** [comment]
test this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 1]
```diff
@@ -936,7 +936,39 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder {
...
+            provider = conf.defaultDataSourceName,
+            temporary = false,
+            partitionColumns = tableDesc.partitionColumnNames.toArray,
+            bucketSpec = None,
+            mode = mode,
+            options = optionsWithPath,
+            q
+          )
+        } else {
+          CreateTableAsSelectLogicalPlan(tableDesc, q, ifNotExists)
```
Comment (ericl) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: Should this one also be renamed to `CreateHiveTableAsSelectLogicalPlan`?

Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: Looking at its implementation, it is not hive-specific. So, seems fine to leave it as is.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala [hunk 2]
```diff
@@ -447,52 +447,20 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
...
 
-        val desc = table.copy(schema = schema)
-
-        if (sessionState.convertCTAS && table.storage.serde.isEmpty) {
-          // Do the conversion when spark.sql.hive.convertCTAS is true and the query
-          // does not specify any storage format (file format and storage handler).
-          if (table.identifier.database.isDefined) {
-            throw new AnalysisException(
-              "Cannot specify database name in a CTAS statement " +
-                "when spark.sql.hive.convertCTAS is set to true.")
```
Comment (ericl) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: Should we keep this restriction?

**Comment by ericl** [comment]
looks good, just have a couple questions

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala (hunk 1 above)
Comment (ericl) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: Seems like `HiveMetastoreCatalog` is the only user though.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 3]
```diff
@@ -936,7 +936,39 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder {
...
+            "Please use the syntax of \"CREATE TABLE tableName USING dataSource " +
+            "OPTIONS (...) PARTITIONED BY ...\" to create a partitioned table through a " +
+            "CTAS statement."
+          throw operationNotAllowed(errorMessage, ctx)
+        }
+
+        val hasStorageProperties = (ctx.createFileFormat != null) || (ctx.rowFormat != null)
+        if (conf.convertCTAS && !hasStorageProperties) {
+          val mode = if (ifNotExists) SaveMode.Ignore else SaveMode.ErrorIfExists
+          val options = rowStorage.serdeProperties ++ fileStorage.serdeProperties
```
Comment (andrewor14) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: I think these will always be empty if we've reached here, no?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionState.scala [hunk 4]
```diff
@@ -139,22 +139,6 @@ private[hive] class HiveSessionState(sparkSession: SparkSession)
...
-   * When true, a table created by a Hive CTAS statement (no USING clause) will be
-   * converted to a data source table, using the data source set by spark.sql.sources.default.
-   * The table in CTAS statement will be converted when it meets any of the following conditions:
-   *   - The CTAS does not specify any of a SerDe (ROW FORMAT SERDE), a File Format (STORED AS), or
-   *     a Storage Handler (STORED BY), and the value of hive.default.fileformat in hive-site.xml
-   *     is either TextFile or SequenceFile.
-   *   - The CTAS statement specifies TextFile (STORED AS TEXTFILE) as the file format and no SerDe
-   *     is specified (no ROW FORMAT SERDE clause).
-   *   - The CTAS statement specifies SequenceFile (STORED AS SEQUENCEFILE) as the file format
-   *     and no SerDe is specified (no ROW FORMAT SERDE clause).
```
Comment (andrewor14) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionState.scala]: should we keep this chunk of documentation?

**Comment by andrewor14** [comment]
LGTM, minor comments only.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala (hunk 2 above)
Comment (yhuai) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: Seems we do support having db name, right?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionState.scala (hunk 4 above)
Comment (yhuai) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionState.scala]: Yea, the useful part is at https://github.com/apache/spark/pull/13386/files#diff-32bb9518401c0948c5ea19377b5069abR313. Other parts are not needed.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala (hunk 3 above)
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: yea

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala (hunk 1 above)
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: ok. Let me change it

**Comment by andrewor14** [comment]
Merging into master 2.0

**Comment by gatorsmile** [comment]
Just realized this PR introduced the original changes. Could you also review my PR: https://github.com/apache/spark/pull/13907?

When users create table as query with `STORED AS` or `ROW FORMAT` and `spark.sql.hive.convertCTAS` is set to true, we do not convert them to data source tables. I am wondering whether we still can convert the tables specified in parquet and orc formats to data source tables? Thanks!

**Pull request closed without merging**